### PR TITLE
Param Navigation

### PR DESCRIPTION
**Pull request by brauliodiez**

### CODE REVIEW

File: 08 ParamNavigation/src/components/app.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import { Router, Route, IndexRoute, Link, IndexLink, browserHistory, hashHistory  } from 'react-router'
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/app.tsx]: Unneeded import

File: 08 ParamNavigation/src/components/app.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/app.tsx]: Unneeded import

File: 08 ParamNavigation/src/components/app.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import { Router, Route, IndexRoute, Link, IndexLink, browserHistory, hashHistory  } from 'react-router'
+
+import Header from './common/header'
+import aboutPage from './about/aboutPage';
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/app.tsx]: Unneeded import

File: 08 ParamNavigation/src/components/app.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import membersPage from './members/membersPage';
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/app.tsx]: Unneeded import

File: 08 ParamNavigation/src/components/app.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+
+interface Props {
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/app.tsx]: You can extends from React.Props that it has children, key and ref properties

```
interface Props extends React.Props<App>{

}
```

File: 08 ParamNavigation/src/components/member/memberForm.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  member : MemberEntity
+  onChange : (event:any) => any;
+  onSave : (event:any) => any;
+  errors: any;
+}
+
+interface State {
+}
+
+export default class memberForm extends React.Component<Props, State> {
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberForm.tsx]: MemberForm

File: 08 ParamNavigation/src/components/member/memberForm.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  constructor(props : Props){
+        super(props);
+  }
+
+
+   public render() {
+       return (
+			    <form>
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberForm.tsx]: Indentation

File: 08 ParamNavigation/src/components/member/memberPage.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+interface State {
+  member : MemberEntity
+  ,errors: any
+  ,dirty  : boolean
+}
+
+// Nice tsx guide: https://github.com/Microsoft/TypeScript/wiki/JSX
+export default class memberPage extends React.Component<Props, State> {
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: rename to MemberPage

File: 08 ParamNavigation/src/components/member/memberPage.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
+import * as React from 'react';
+import { hashHistory } from 'react-router'
+import * as toastr from 'toastr';
+import MemberEntity from './../../api/memberEntity'
+import MemberForm from './memberForm';
+import MemberAPI from '../../api/memberAPI';
+//import * as ObjectAssign from 'object-assign';
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: remove comment

File: 08 ParamNavigation/src/components/member/memberPage.tsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      return this.setState(newState);
+
+    }
+  }
+
+  // on any update on the form this function will be called
+  setMemberState(event) {
+    // https://www.npmjs.com/package/object-assign
+    //var newState : State = objectAssign({}, this.state, {dirty: true});
+    //this.setState(newState);
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: remove comments

File: 08 ParamNavigation/src/components/member/memberPage.tsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        super(props);
+        // set initial state
+        this.state = {
+                      member: new MemberEntity()
+                      ,errors : {}
+                      ,dirty : false
+                     };
+  }
+
+  componentWillMount() {
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: componentWillMount() or componentDidMount()?

> **Mounting: componentWillMount**
> void componentWillMount()
> Invoked once, both on the client and server, immediately before the initial rendering occurs. If you call setState within this method, render() will see the updated state and will be executed only once despite the state change.


> **Mounting: componentDidMount #**
> void componentDidMount()
> Invoked once, only on the client (not on the server), immediately after the initial rendering occurs. At this point in the lifecycle, you can access any refs to your children (e.g., to access the underlying DOM representation). The componentDidMount() method of child components is invoked before that of parent components.
> 
> If you want to integrate with other JavaScript frameworks, set timers using setTimeout or setInterval, or send AJAX requests, perform those operations in this method.

[https://facebook.github.io/react/docs/component-specs.html](https://facebook.github.io/react/docs/component-specs.html)

I think the key is in 

> or send AJAX requests

File: 08 ParamNavigation/src/components/app.tsx (hunk 1 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/app.tsx]: done

File: 08 ParamNavigation/src/components/app.tsx (hunk 2 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/app.tsx]: done

File: 08 ParamNavigation/src/components/member/memberPage.tsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+  event.preventDefault();
+
+  if(!this.memberFormIsValid()) {
+    return;
+  }
+
+  var memberAPI : MemberAPI = new MemberAPI();
+  memberAPI.saveAuthor(this.state.member);
+
+  var newState : State = objectAssign({}, this.state, {dirty: true});
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: Why you have to set dirty?

File: 08 ParamNavigation/src/components/app.tsx (hunk 3 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/app.tsx]: done

File: 08 ParamNavigation/src/components/app.tsx (hunk 4 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/app.tsx]: done

File: 08 ParamNavigation/src/components/app.tsx (hunk 5 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/app.tsx]: cool !

File: 08 ParamNavigation/src/components/member/memberForm.tsx (hunk 6 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/member/memberForm.tsx]: Done

File: 08 ParamNavigation/src/components/members/membersPage.tsx [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  constructor(props : Props){
+        super(props);
+        // set initial state
+        this.state = {members: []};
+  }
+
+
+   // Standard react lifecycle function:
+   // https://facebook.github.io/react/docs/component-specs.html
+   public componentWillMount() {
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/members/membersPage.tsx]: Same like before in memberPage.tsx

File: 08 ParamNavigation/src/components/member/memberForm.tsx (hunk 7 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/member/memberForm.tsx]: Done

File: 08 ParamNavigation/src/components/member/memberPage.tsx (hunk 8 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: Done

File: 08 ParamNavigation/src/components/members/membersPage.tsx [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+   public render() {
+
+       return (
+        <div className="row">
+          <h2> Members Page</h2>
+          <Link to="/member">New Member</Link>
+          <table className="table">
+            <thead>
+              <th>
```
Comment (Nasdan) [inline 08 ParamNavigation/src/components/members/membersPage.tsx]: Console throw warning if you don't wrap <th> inside a <tr>

```
<thead>
   <tr>
      <th>
      </th>
  </tr>
</thead>
```

File: 08 ParamNavigation/src/components/member/memberPage.tsx (hunk 9 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: Done

File: 08 ParamNavigation/src/components/member/memberPage.tsx (hunk 10 above)
Comment (brauliodiez) [inline 08 ParamNavigation/src/components/member/memberPage.tsx]: Done

**Pull request merged**